internal/helm/dependency_manager.go

```diff
@@ -28,146 +28,287 @@ import (
 	"github.com/Masterminds/semver/v3"
 	securejoin "github.com/cyphar/filepath-securejoin"
 	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
 	helmchart "helm.sh/helm/v3/pkg/chart"
 	"helm.sh/helm/v3/pkg/chart/loader"
 )
 
-// DependencyWithRepository is a container for a Helm chart dependency
-// and its respective repository.
-type DependencyWithRepository struct {
-	// Dependency holds the reference to a chart.Chart dependency.
-	Dependency *helmchart.Dependency
-	// Repository is the ChartRepository the dependency should be
-	// available at and can be downloaded from. If there is none,
-	// a local ('file://') dependency is assumed.
-	Repository *ChartRepository
-}
+// GetChartRepositoryCallback must return a ChartRepository for the URL,
+// or an error describing why it could not be returned.
+type GetChartRepositoryCallback func(url string) (*ChartRepository, error)
 
-// DependencyManager manages dependencies for a Helm chart.
+// DependencyManager manages dependencies for a Helm chart, downloading
+// only those that are missing from the chart it holds.
 type DependencyManager struct {
-	// WorkingDir is the chroot path for dependency manager operations,
+	// chart contains the chart.Chart from the path.
+	chart *helmchart.Chart
+
+	// baseDir is the chroot path for dependency manager operations,
 	// Dependencies that hold a local (relative) path reference are not
 	// allowed to traverse outside this directory.
-	WorkingDir string
-	// ChartPath is the path of the Chart relative to the WorkingDir,
-	// the combination of the WorkingDir and ChartPath is used to
+	baseDir string
+
+	// path is the path of the chart relative to the baseDir,
+	// the combination of the baseDir and path is used to
 	// determine the absolute path of a local dependency.
-	ChartPath string
-	// Chart holds the loaded chart.Chart from the ChartPath.
-	Chart *helmchart.Chart
-	// Dependencies contains a list of dependencies, and the respective
-	// repository the dependency can be found at.
-	Dependencies []*DependencyWithRepository
+	path string
+
+	// repositories contains a map of ChartRepository indexed by their
+	// normalized URL. It is used as a lookup table for missing
+	// dependencies.
+	repositories map[string]*ChartRepository
 
+	// getChartRepositoryCallback can be set to an on-demand get
+	// callback which returned result is cached to repositories.
+	getChartRepositoryCallback GetChartRepositoryCallback
+
+	// workers is the number of concurrent chart-add operations during
+	// Build. Defaults to 1 (non-concurrent).
+	workers int64
```

Comment:
Are you planning to make this configurable for the user at some point?
I assume this does affect memory utilization so I agree it should be set cautiously to a lower value.

Reply:
I made it configurable because it used to be parallel and I didn't want to loose the functionality (in code).

However, based on the user feedback and having a closer look at the code, it seems to be little beneficial to do these things in parallel as network operations should generally be quick, while memory usage is expensive.